### PR TITLE
Add LUT1DEffect

### LINKED ISSUE #368: Support for 1D LUTs
For [three-filmic](https://github.com/donmccurdy/three-filmic), I needed a small 1D "shaper" LUT in addition to the larger 3D LUT. The 1D version is really simple, implemented here as [LUT1DEffect.ts](https://github.com/donmccurdy/three-filmic/blob/main/src/effects/LUT1DEffect.ts). Would you be interested in having support for 1D LUTs in this repository? If so, I'd be happy to make a PR. Should it be part of LUTEffect, or separate the two as (for example) LUT1DEffect and LUT3DEffect?

Related — I found that LUTCubeLoader does not currently support 1D `.cube` LUTs. That'd be pretty easy to add as well, if we're interested. I wasn't sure why LUTCubeLoader is forked from the three.js version though — is there an important difference between the two? For bandwidth reasons, I'm planning to eventually switch three-filmic from plaintex `.cube` LUTs to compressed `.ktx2` data textures.

### Additional context

- https://github.com/pmndrs/postprocessing/issues/124
- https://github.com/pmndrs/postprocessing/issues/231

### PR DESCRIPTION
**Pull request by donmccurdy**
Changes:

- Fixes #368
- Adds LUT1DEffect
- Renames LUTEffect → LUT3DEffect (with deprecated alias)
- ~~Update DataTexture3D → Data3DTexture~~
- Include 'inputEncoding' as a LUT3DEffect constructor parameter

Questions:

- [x] ~~I'm not sure how to update `postprocessing.d.ts`, which prevents me from testing the change in my own project — is that automated?~~
- [x] ~~I've omitted the input/output encoding steps in LUT1DEffect for now. Based on discussion in #370 it sounded like that might eventually be an Effect-level thing rather than living in LUT1DEffect and LUT3DEffect? It doesn't affect my current use case, but let me know what you'd prefer here.~~

### CODE REVIEW

**Comment by vanruesc** [comment]
> I'm not sure how to update `postprocessing.d.ts`, which prevents me from testing the change in my own project — is that automated?

@superguigui used a [tool](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html) for that but he had to adjust the result by hand. I've also updated some things manually by searching through `postprocessing.d.ts`. Feel free to ignore the types - I can adjust that part later.

> I've omitted the input/output encoding steps in LUT1DEffect for now.

That's fine. `Effect.processColorSpace` will indeed add that functionality.

**Comment by donmccurdy** [comment]
Ok, thanks! I'll work out how to test this change in my project locally (`yarn link` is behaving weirdly with my build tool for some reason...) and then un-draft the PR after that.

**Comment by donmccurdy** [comment]
Confirmed this change is working in three-filmic. 

Here's a simple example of a neutral 1D LUT:

```js
const NEUTRAL_LUT_1D = new DataTexture(
	new Float32Array([0, 0.25, 0.5, 0.75, 1]),
	5,
	1,
	RedFormat,
	FloatType,
	UVMapping,
	ClampToEdgeWrapping,
	ClampToEdgeWrapping,
	LinearFilter,
	LinearFilter
);
NEUTRAL_LUT_1D.name = 'neutral1D';
NEUTRAL_LUT_1D.needsUpdate = true;
```

One important thing to note — [WebGL 2 requires THREE.RGBAFormat](https://github.com/mrdoob/three.js/blob/3b08a8fc32298b00a137a4e5470ffecfe7f0fec4/src/renderers/webgl/WebGLTextures.js#L1817-L1823) when using `texture.encoding = THREE.sRGBEncoding`. Since these 1D textures use THREE.RedFormat, they must use THREE.LinearEncoding instead. In my case the inputs are a custom log encoding, and the outputs of the 1D LUT are known to be sRGB (although I could also change that). I'm manually disabling `this.fullscreenMaterial.encodeOutput` to avoid double-encoding. This may be a bit hard to handle with just `Effect.processColorSpace`, as LUTs can change the data's color space. 😕

**Review by vanruesc — CHANGES_REQUESTED** [review]

**Comment by vanruesc** [comment]
> One important thing to note — [WebGL 2 requires THREE.RGBAFormat](https://github.com/mrdoob/three.js/blob/3b08a8fc32298b00a137a4e5470ffecfe7f0fec4/src/renderers/webgl/WebGLTextures.js#L1817-L1823) when using texture.encoding = THREE.sRGBEncoding. Since these 1D textures use THREE.RedFormat, they must use THREE.LinearEncoding instead.

It's my understanding that most LUTs themselves use `LinearEncoding` and that their color values should be used as-is. If a user wants to convert a 1D LUT to sRGB on the fly via `sRGBEncoding`, then they could always convert it to `RGBAFormat` by duplicating the values for each color channel.

>  In my case the inputs are a custom log encoding, and the outputs of the 1D LUT are known to be sRGB (although I could also change that). I'm manually disabling this.fullscreenMaterial.encodeOutput to avoid double-encoding. This may be a bit hard to handle with just Effect.processColorSpace, as LUTs can change the data's color space.

> We could also just ignore the LUT1D special case here, and I'll modify my LUTs to output Linear-sRGB, so the output encoding works as expected.

Maybe the approach with `inputColorSpace` & `outputColorSpace` would be better after all because that would explicitly inform the `EffectPass` about the output color space of each effect. That way we don't leave room for assumptions.

File: src/effects/LUT3DEffect.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import {
-	DataTexture3D,
+	Data3DTexture,
```
[the remark this answers is not in the document]
Comment (donmccurdy) [reply src/effects/LUT3DEffect.js]: Reverted ✅

File: src/images/textures/LookupTexture.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ import {
 	Color,
 	ClampToEdgeWrapping,
 	DataTexture,
-	DataTexture3D,
+	Data3DTexture,
```
[the remark this answers is not in the document]
Comment (donmccurdy) [reply src/images/textures/LookupTexture.js]: Reverted ✅

**Comment by donmccurdy** [comment]
Reverted the DataTexture3D → Data3DTexture change.

I'm seeing the same error about LookupTexture using sRGB now too, because it's FloatType and not UnsignedIntType. But agreed, I'd be surprised if most users are encoding their display transform in the LUT, it seems fine to assume Linear-sRGB or no color space instead.

The `inputColorSpace` and `outputColorSpace` properties make sense then. These should probably be readonly, since it's not the right way for users to change the output color space of an effects stack?

**Review by vanruesc — APPROVED** [review]

**Comment by vanruesc** [comment]
Thanks!

**Pull request merged**